### ISSUE TITLE
Changing setuid permissions for security reasons in some binaries installed through RPM causes the HasModifiedFiles check in Preflight 1.6.9 to fail

### ISSUE THREAD

**Issue by ramperher** [redhat-openshift-ecosystem/openshift-preflight#1025]
## Bug Description
Changing setuid permissions for security reasons in some binaries installed through RPM causes the HasModifiedFiles check in Preflight 1.6.9 to fail

## Version and Command Invocation
Preflight 1.6.9

## Steps to Reproduce:

1) Build an image quay.io/raperez/setuid-example:v1.0.0 using the Dockerfile below:

```
FROM registry.access.redhat.com/ubi8/ubi-minimal

RUN microdnf update --setopt=tsflags=nodocs -y && \
    microdnf install --setopt=tsflags=nodocs -y nc shadow-utils tar zip unzip findutils procps-ng && \
    microdnf clean all

RUN find / -xdev -perm -4000 -exec chmod ug-s {} +

CMD [ "sleep", "3600"]
```

2) Run the Preflight check and observe the HasModifiedFiles check fail:

```
$ podman run --privileged -e PFLT_LOGLEVEL=trace -e PFLT_DOCKERCONFIG=/opt/config.json -v ./config.json:/opt/config.json quay.io/opdev/preflight:1.6.9 check container quay.io/raperez/setuid-example:v1.0.0
...
time="2023-06-23T11:43:40Z" level=debug msg="found disallowed modification in layer" file=usr/bin/newgrp layer="sha256:1db831829344088d5d1cc23954bf8cc3d7f8d350c239e65a4deebd0fd7b8d0dd"
time="2023-06-23T11:43:40Z" level=debug msg="found disallowed modification in layer" file=usr/bin/chage layer="sha256:1db831829344088d5d1cc23954bf8cc3d7f8d350c239e65a4deebd0fd7b8d0dd"
time="2023-06-23T11:43:40Z" level=debug msg="found disallowed modification in layer" file=usr/bin/gpasswd layer="sha256:1db831829344088d5d1cc23954bf8cc3d7f8d350c239e65a4deebd0fd7b8d0dd"
time="2023-06-23T11:43:40Z" level=info msg="check completed" check=HasModifiedFiles result=FAILED
```

It's because we have changed the setuid permissions of these three binaries. If we check the permissions before changing them, we'll see they have it:

```
[root@f83a39b0d719 /]# ls /usr/bin/ | grep chage
chage
[root@f83a39b0d719 /]# ls /usr/bin/ | grep gpasswd
gpasswd
[root@f83a39b0d719 /]# ls /usr/bin/ | grep newgrp
newgrp
[root@f83a39b0d719 /]# find / -xdev -perm -4000
/usr/bin/chage
/usr/bin/gpasswd
/usr/bin/newgrp
```

3) Confirm that the issue is caused because of changing the setuid permissions by building a new image that does not change any permission, called quay.io/raperez/setuid-fix-example:v1.0.0:

```
FROM registry.access.redhat.com/ubi8/ubi-minimal

RUN microdnf update --setopt=tsflags=nodocs -y && \
    microdnf install --setopt=tsflags=nodocs -y nc shadow-utils tar zip unzip findutils procps-ng && \
    microdnf clean all

CMD [ "sleep", "3600"]
```

This image passes the HasModifiedFiles check:

```
$ podman run --privileged -e PFLT_LOGLEVEL=trace -e PFLT_DOCKERCONFIG=/opt/config.json -v ./config.json:/opt/config.json quay.io/opdev/preflight:1.6.9 check container quay.io/raperez/setuid-fix-example:v1.0.0
...
time="2023-06-23T11:42:35Z" level=info msg="check completed" check=HasModifiedFiles result=PASSED
```

## Expected Result

The removal of setuid permissions in the container's filesystem is an action that is required to mitigate some kind of attacks, such as some open-source projects like [Elasticsearch](https://github.com/elastic/elasticsearch/blob/main/distribution/docker/src/docker/Dockerfile#L211-L212) report in their Dockerfiles.

This is an action that can be required in some cases, but as it implies to modify the default permissions of the binaries present in the container, then the RPM packages downloaded would change consequently. Anyway, as it's a recommended step, it's just to ensure the correct performance of the workload, so should not impact in certification.

## Actual Result

check=HasModifiedFiles result=FAILED

## Additional Context

The instructions for the Dockerfile we have provided are just extracted from this [Dockerfile](https://github.com/elastic/elasticsearch/blob/main/distribution/docker/src/docker/Dockerfile) from that repository. We are trying to certify a container image from a partner that is using these instructions as base, and that's the only issue we currently have with that image; the rest of checks are correct.

Note that there's a [post](https://discuss.elastic.co/t/elasticsearch-distribution-docker-src-docker-dockerfile-regarding-the-absence-of-files-with-setuid/336378) opened in Elasticsearch support regarding this.

cc: @bcrochet @acornett21 @tkrishtop

**Comment by vladbutusinakeysight** [comment]
Hi,
This a good read regarding the issue presented.
https://docs.oracle.com/cd/E19253-01/816-4557/secfile-69/index.html

I believe that removing the setuit and setguid permission bits in the base image (UBI), for the files that are found as modified in our built image (also per elastic source [dockerfile](https://github.com/elastic/elasticsearch/blob/main/distribution/docker/src/docker/Dockerfile#L211-L212)), is actually better for the base image, as it might prevent some security risks.
We would then remove the find and permissions removal line ...
Thanks.

**Comment by bcrochet** [comment]
My inclination is to say that this is a legitimate fail on this check. Changing the permissions on a RH-installed file is not recommended. 

When you say that the 'action is required', I'm curious where this "requirement" comes from?

**Comment by ramperher** [comment]
Hi @bcrochet . This requirement comes from Elasticsearch open-source community, I just included the link where they justify why they need to change setuid permissions in the description, putting it [here](https://github.com/elastic/elasticsearch/blob/main/distribution/docker/src/docker/Dockerfile#L211-L212) again.

**Comment by acornett21** [comment]
How I see things, the post from elasticsearch is a `nice to have` not a `requirement to make work`, which are two drastically different things in my opinion. We are still bound by the [certification policies ](https://access.redhat.com/documentation/en-us/red_hat_software_certification/8.64/html/red_hat_openshift_software_certification_policy_guide/assembly-requirements-for-container-images_openshift-sw-cert-policy-introduction), and changing `setuid` violates these polices.

**Comment by bcrochet** [comment]
> Hi @bcrochet . This requirement comes from Elasticsearch open-source community, I just included the link where they justify why they need to change setuid permissions in the description, putting it [here](https://github.com/elastic/elasticsearch/blob/main/distribution/docker/src/docker/Dockerfile#L211-L212) again.

I read the link. It still doesn't justify a change in preflight or the policy.

The last information I can find on this 'stack clash' vulnerability is circa 2017. And there are no known remote exploits. Given that we are talking about containers, I just find this a bit extreme, and frankly, unnecessary.

 https://access.redhat.com/security/vulnerabilities/stackguard

**Comment by bcrochet** [comment]
> Hi, This a good read regarding the issue presented. https://docs.oracle.com/cd/E19253-01/816-4557/secfile-69/index.html
> 
> I believe that removing the setuit and setguid permission bits in the base image (UBI), for the files that are found as modified in our built image (also per elastic source [dockerfile](https://github.com/elastic/elasticsearch/blob/main/distribution/docker/src/docker/Dockerfile#L222)), is actually better for the base image, as it might prevent some security risks. We would then remove the find and permissions removal line ... Thanks.

Then getting it changed in UBI would be the place to affect this change, not preflight.

**Comment by tkrishtop** [comment]
> Then getting it changed in UBI would be the place to affect this change, not preflight.

hi @bcrochet, thank you for checking! What solution do you propose for the partner? They have already checked with their security office, and they need to follow Elasticsearch recommendations. However, they also want to obtain RH certification for their image.

**Comment by lslebodn** [comment]
```
# 3. Ensure that there are no files with setuid or setgid, in order to mitigate "stackclash" attacks.
#    We've already run this in previous layers so it ought to be a no-op.
```
@ramperher  files with setuid or setgid are not useful if you run container with dropped permissions.
e.g. `docker run --security-opt=no-new-privileges`
or
```
[root@localhost ]# /usr/bin/docker run -ti  --rm registry.access.redhat.com/ubi9/ubi bash -c 'id && su --shell=/bin/sh --command id dbus'
uid=0(root) gid=0(root) groups=0(root)
uid=81(dbus) gid=81(dbus) groups=81(dbus)

[root@localhost ]# /usr/bin/docker run -ti --cap-drop=SETUID --cap-drop=SETUID --rm registry.access.redhat.com/ubi9/ubi bash -c 'ls -l /bin/su && id && su --shell=/bin/sh --command id dbus'
-rwsr-xr-x. 1 root root 56936 Mar 29 10:46 /bin/su
uid=0(root) gid=0(root) groups=0(root)
su: cannot set user id: Operation not permitted
```

**Comment by ramperher** [comment]
> ```
> # 3. Ensure that there are no files with setuid or setgid, in order to mitigate "stackclash" attacks.
> #    We've already run this in previous layers so it ought to be a no-op.
> ```
> 
> @ramperher files with setuid or setgid are not useful if you run container with dropped permissions. e.g. `docker run --security-opt=no-new-privileges` or
> 
> ```
> [root@localhost ]# /usr/bin/docker run -ti  --rm registry.access.redhat.com/ubi9/ubi bash -c 'id && su --shell=/bin/sh --command id dbus'
> uid=0(root) gid=0(root) groups=0(root)
> uid=81(dbus) gid=81(dbus) groups=81(dbus)
> 
> [root@localhost ]# /usr/bin/docker run -ti --cap-drop=SETUID --cap-drop=SETUID --rm registry.access.redhat.com/ubi9/ubi bash -c 'ls -l /bin/su && id && su --shell=/bin/sh --command id dbus'
> -rwsr-xr-x. 1 root root 56936 Mar 29 10:46 /bin/su
> uid=0(root) gid=0(root) groups=0(root)
> su: cannot set user id: Operation not permitted
> ```

Thanks for the suggestion @lslebodn ! Anyway, this does not affect to the result reported here.

**Comment by lslebodn** [comment]
> Thanks for the suggestion @lslebodn ! Anyway, this does not affect to the result reported here.

It was already mentioned in previous comments: **Changing the permissions on a RH-installed file is not recommended**. 
In another words, the reproducer is not ideal
```
RUN find / -xdev -perm -4000 -exec chmod ug-s {} +
```

I know initial report is oversimplified and not real case. 
But one can use container engine runtime options to drop capabilities/privileges even without modification to base image.
Which is very often huge simplification

@ramperher I just wanted to provider a better and more secure way how to mitigate theoretical issues mentioned by elasticsearch.

**Comment by ramperher** [comment]
> > Thanks for the suggestion @lslebodn ! Anyway, this does not affect to the result reported here.
> 
> It was already mentioned in previous comments: **Changing the permissions on a RH-installed file is not recommended**. In another words, the reproducer is not ideal
> 
> ```
> RUN find / -xdev -perm -4000 -exec chmod ug-s {} +
> ```
> 
> I know initial report is oversimplified and not real case. But one can use container engine runtime options to drop capabilities/privileges even without modification to base image. Which is very often huge simplification
> 
> @ramperher I just wanted to provider a better and more secure way how to mitigate theoretical issues mentioned by elasticsearch.

Yes, I got that point, the problem is that, here, the commands came from Elasticsearch project, so we're just testing the same commands that are defined in their Dockerfile and that are causing the issue with HasModifiedFiles.

By the way, as @tkrishtop commented last week, @bcrochet , do you have any updates regarding what to comment with the partner to unblock this?

**Comment by acornett21** [comment]
@ramperher We will not be changing preflight to accommodate this partner. The partner would have to seek other means to get their container(s) certified, ie an exception process with the business, if they allow.

**Comment by bcrochet** [comment]
Closing as WONTFIX.
